cmd/ctl/pkg/check/api/api.go

```diff
@@ -92,41 +89,40 @@ func NewCmdCheckApi(ctx context.Context, ioStreams genericclioptions.IOStreams)
 		Use:   "api",
 		Short: "Check if the cert-manager API is ready",
 		Long:  checkApiDesc,
-		RunE: func(cmd *cobra.Command, args []string) error {
-			if err := o.Complete(); err != nil {
-				return err
-			}
-			o.Run(ctx)
-			return nil
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(o.Complete())
+			cmdutil.CheckErr(o.Run(ctx))
 		},
-		SilenceUsage:  true,
-		SilenceErrors: true,
 	}
 	cmd.Flags().DurationVar(&o.Wait, "wait", 0, "Wait until the cert-manager API is ready (default 0s = poll once)")
 	cmd.Flags().DurationVar(&o.Interval, "interval", 5*time.Second, "Time between checks when waiting, must include unit, e.g. 1m or 10m")
-	cmd.Flags().BoolVarP(&o.Verbose, "verbose", "v", false, "Print detailed error messages")
 
 	o.Factory = factory.New(ctx, cmd)
 
 	return cmd
 }
 
 // Run executes check api command
-func (o *Options) Run(ctx context.Context) {
-	if !o.Verbose {
-		log.SetFlags(0) // Disable prefixing logs with timestamps.
-	}
-	log.SetOutput(o.ErrOut) // Log all intermediate errors to stderr
+func (o *Options) Run(ctx context.Context) error {
+	log := logf.FromContext(ctx, "checkAPI")
 
 	start := time.Now()
-	pollErr := wait.PollUntilContextCancel(ctx, o.Interval, false, func(ctx context.Context) (bool, error) {
+	var lastError error
+	pollErr := wait.PollUntilContextCancel(ctx, o.Interval, true, func(ctx context.Context) (bool, error) {
 		if err := o.APIChecker.Check(ctx); err != nil {
-			if !o.Verbose && errors.Unwrap(err) != nil {
-				err = errors.Unwrap(err)
+			simpleError := cmapichecker.TranslateToSimpleError(err)
+			if simpleError != nil {
+				if !log.V(2).Enabled() {
+					log.Error(simpleError, "Not ready")
+				} else {
+					log.Error(simpleError, "Not ready", "underlyingError", err)
+				}
+				lastError = simpleError
+			} else {
+				log.Error(err, "Not ready")
+				lastError = err
```

Comment:
It doesn't look OK to me to handle the error but still return it afterwards. There is no point in displaying the error at the end if we already displayed it.

Also, why do we print a message to show the progress? The `--wait` parameter already shows the maximum wait time, so I don't understand the purpose of showing intermediate log lines. `kubectl wait` doesn't either:

```console
$ kubectl wait deploy -n cert-manager cert-manager-webhook --for=condition=Ready --timeout=30s
error: timed out waiting for the condition on deployments/cert-manager-webhook
```

**suggestion:** don't show intermediate log lines. If people really want to have some intermediate output (e.g., to prevent CI's streaming from getting stale), they can use `-v=6`.

Reply:
I updated the PR, now we only log intermediate steps on level 2 (= when the `-v` flag is set)